lib/xml/reader.dart

```diff
@@ -141,3 +144,147 @@ class XmlReader {
     return context.success(null, context.position + 1);
   }
 }
+
+/// A push based XmlReader interface, intended to be similar to .NET's XmlReader.
+class XmlTextReader {
+  /// Creates a new reader for `input`.
+  /// 
+  /// Setting `ignoreWhitespace` to false will cause text nodes 
+  XmlTextReader(String input, {this.ignoreWhitespace = true}) {
+    _result = Success(input, 0, null);
+    _depth = 0;
+    _eof = false;
+  }
+
+  /// If true, will ignore `XmlNodeType.TEXT` and when it is composed of whitespace.
+  bool ignoreWhitespace;
+
+  /// Parsing context.
+  Result _result;
+
+  /// The [XmlNodeType] of the current position of the reader.
+  XmlNodeType get nodeType => _nodeType;
+  XmlNodeType _nodeType;
+
+  /// The [XmlName] of the current node of the reader.
+  XmlName get name => _name;
+  XmlName _name;
+
+  /// The value of the current node of the reader.
+  ///
+  /// The value will have
+  String get value => _value;
+  String _value;
+
+  /// Will return true if `nodeType == XmlNodeType.ELEMENT` and the element is self closing,
+  /// e.g. `<element />`.
+  bool get isEmptyElement => _isEmptyElement;
+  bool _isEmptyElement;
+
+  /// The zero based depth of the reader.
+  int get depth => _depth;
+  int _depth;
+
+  /// True if the reader is positioned at the end of the buffer.
+  bool get eof => _eof;
+  bool _eof;
+
+  /// The `List<XmlAttribute>` of the current element (if `nodeType == XmlNodeType.ELEMENT`).
+  UnmodifiableListView<XmlAttribute> get attributes =>
+      UnmodifiableListView<XmlAttribute>(_attributes);
+  List<XmlAttribute> _attributes;
+
+  /// Advances the reader to the next readable position.  Returns true if more data remains, false if not.
+  bool read() {
+    if (_result.position > _result.buffer.length) {
+      _eof = true;
+      return false;
+    }
+
+    _result = _parseEvent(_result);
+    return (_result.isSuccess);
+  }
+
+  Result _parseEvent(Result context) {
```

Comment:
I'd prefer, if we'd somehow reuse the implementation of `XmlReader#_parseEvent`. If this is not possible, we'd need to extract the common code somehow.

Reply:
They work pretty differently.  I'll see if I can come up with something, but most of the commonality is in the `_characterData`, `_elementStart`, etc.